### PR TITLE
NET: Yahya/5147-deduplicate-channels-part 1

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR presents the effors on refactoring the channel IDs from Golang native string type, to a type definition. We are doing this for better encapsulation and representation of `Channel`s as well as to prepare the ground for building up the deduplication mechanism of channels in the overlay topology of the network layer. Defining the `Channel` as a type allows us to later refactor its representation easier and transparent without touching any other domain. 

In the new version of network layer, `Channel` is an abstract and virtual concept at the _application_ level that corresponds to the _topic_ in libp2p. Nodes can subscribe or unsubscribe to a channel. While the nodes are subscribing to a channel, they can send and receive epidemic messages to/from other nodes subscribed to that channel.

### CODE REVIEW

File: engine/channels.go [hunk 1]
```diff
@@ -156,23 +157,23 @@ func IsClusterChannelID(channelID string) (string, bool) {
 
 // FullyQualifiedChannelName returns the unique channel name made up of channel name string suffixed with root block id
 // The root block id is used to prevent cross talks between nodes on different sporks
-func FullyQualifiedChannelName(channelID string, rootBlockID string) string {
+func FullyQualifiedChannelName(channel network.Channel, rootBlockID string) string {
```
Comment (vishalchangrani) [inline engine/channels.go:160]: We could rename this function to `TopicFromChannel`. It is used to suffix the channel name with the root block id.
If you look at all the usages of this function, the callers store the return value into a variable called `topic` :)
That way we can also get rid of this notion of `FullyQualifiedChannelName` and only gave `channels` at the upper layer and `topic` at the lower layer.

**Review by vishalchangrani — APPROVED** [review]
lgtm

File: network/p2p/network.go [hunk 2]
```diff
@@ -109,18 +109,18 @@ func (n *Network) Done() <-chan struct{} {
 // Register will register the given engine with the given unique engine engineID,
 // returning a conduit to directly submit messages to the message bus of the
 // engine.
-func (n *Network) Register(channelID string, engine network.Engine) (network.Conduit, error) {
-	if _, ok := channels.RolesByChannelID(channelID); !ok {
-		return nil, fmt.Errorf("unknown channel id: %s, should be registered in topic map", channelID)
+func (n *Network) Register(channel network.Channel, engine network.Engine) (network.Conduit, error) {
+	if _, ok := channels.RolesByChannel(channel); !ok {
+		return nil, fmt.Errorf("unknown channel: %s, should be registered in topic map", channel)
 	}
```
Comment (jordanschalm) [inline network/p2p/network.go:115]: We could add a utility function for this for readability
```suggestion
	if !channels.Exists(channel) {	
		return nil, fmt.Errorf("unknown channel: %s", channel)
	}
```

**Review by jordanschalm — APPROVED** [review]
👍 LGTM - nice work

File: engine/channels.go (hunk 1 above)
Comment (yhassanzadeh13) [reply engine/channels.go]: Defined `Topic` as a type, and applied the comment.

File: network/p2p/network.go (hunk 2 above)
Comment (yhassanzadeh13) [reply network/p2p/network.go]: added here https://github.com/onflow/flow-go/pull/289/commits/8b3a21ad501d166b80e44900a018347820e93942

**Pull request merged**